dkg/frost/keygen_output.go

```diff
@@ -0,0 +1,144 @@
+package frost
+
+import (
+	"bytes"
+
+	"github.com/bloxapp/ssv-spec/dkg"
+	"github.com/bloxapp/ssv-spec/types"
+
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/pkg/errors"
+)
+
+// processKeygenOutput function can be executed once all messages from round 2
+// have been received. It reconstructs the Vk using the received VkShares and
+// verifies that it is consistent. The keygen output is then created, including
+// the Validator PK, operator public keys, and the operator's own share, and
+// returned as a Protocol Outcome.
+func (fr *Instance) processKeygenOutput() (finished bool, protocolOutcome *dkg.ProtocolOutcome, err error) {
+
+	if !fr.canProceedThisRound() {
+		return false, nil, nil
+	}
+	fr.state.currentRound = KeygenOutput
+
+	if !fr.needToRunCurrentRound() {
+		return false, nil, nil
+	}
+
+	// verify shares by reconstructing consistent validatorPK over a sliding
+	// window of t (threshold) operators
+	reconstructed, err := fr.verifyShares()
+	if err != nil {
+		return false, nil, errors.Wrap(err, "failed to verify shares")
+	}
+	reconstructedBytes := reconstructed.Serialize()
+
+	// prepare keygen output
+	out := &dkg.KeyGenOutput{
+		Threshold: uint64(fr.instanceParams.threshold),
+	}
+
+	operatorPubKeys := make(map[types.OperatorID]*bls.PublicKey)
+	for _, operatorID := range fr.instanceParams.operators {
+		msg, err := fr.state.msgContainer.GetRound2Msg(operatorID)
+		if err != nil {
+			return false, nil, errors.Wrap(err, "failed to retrieve round2 msg")
+		}
+
+		// set vk and secret share to output for this participant
+		if operatorID == uint32(fr.instanceParams.operatorID) {
+			out.ValidatorPK = msg.Vk
+			sk := &bls.SecretKey{}
+			if err := sk.Deserialize(fr.state.participant.SkShare.Bytes()); err != nil {
+				return false, nil, err
+			}
+			out.Share = sk
+		}
+
+		// set operator public key for all operators
+		pk := &bls.PublicKey{}
+		if err := pk.Deserialize(msg.VkShare); err != nil {
+			return false, nil, err
+		}
+		operatorPubKeys[types.OperatorID(operatorID)] = pk
+	}
+	out.OperatorPubKeys = operatorPubKeys
+
+	// assert validatorPK output from frost is equal to reconstructed Vk
+	if !bytes.Equal(out.ValidatorPK, reconstructedBytes) {
+		return false, nil, errors.New("can't reconstruct to the validator pk")
+	}
+
+	return true, &dkg.ProtocolOutcome{ProtocolOutput: out}, nil
+}
+
+// verifyShares reconstructs Vk over a sliding window of t (threshold) operators
+// and checks if reconstructed Vk is consistent
+func (fr *Instance) verifyShares() (*bls.G1, error) {
+
+	var (
+		quorumStart       = 0
+		prevReconstructed = (*bls.G1)(nil)
+	)
+
+	// Sliding window of quorum 0...threshold until n-threshold...n
```

Comment:
Add the rationale: by using a sliding window, every share is being used at least once, if every reconstruction produces an identical Vk, we can tell all the shares are valid.